src/dbnode/persist/fs/retriever.go

```diff
@@ -311,10 +333,24 @@ func (r *blockRetriever) fetchBatch(
 		return
 	}
 
+	// NB: filter out stream checksum requests, as these are handled seperately.
+	streamReqs := reqs[:0]
 	// Sort the requests by offset into the file before seeking
 	// to ensure all seeks are in ascending order
 	var limitErr error
 	for _, req := range reqs {
+		if req.streamReqType == streamInvalidReq {
+			req.onError(errUnsetRequestType)
+			continue
+		}
+
+		// NB: index checksum queries should ignore limits.
```

Comment:
Maybe need to do this at entry to this function? There is a `r.queryLimits.AnyExceeded()` at the beginning.

Perhaps best to re-arrange to get the seeker, then process any requests that are possibly stream index requests, then resume the normal flow? i.e.
```go
	// Resolve the seeker from the seeker mgr
	seeker, err := seekerMgr.Borrow(shard, blockStart)
	if err != nil {
		for _, req := range reqs {
			req.onError(err)
		}
		return
	}

	defer func() {
		err := seekerMgr.Return(shard, blockStart, seeker)
		if err != nil {
			r.logger.Error("err returning seeker for shard",
				zap.Uint32("shard", shard),
				zap.Int64("blockStart", blockStart.Unix()),
				zap.Error(err),
			)
		}
	}()

	// Rebuild reqs so the wrong reference is less likely to be used later.
	reqsPreProcessing := reqs
	reqs = reqs[:0]

	// Deal with index checksum stream requests first.
	for _, req := range reqsPreProcessing {
		if req.streamReqType == streamInvalidReq {
			req.onError(errUnsetRequestType)
			continue
		}

		if req.streamReqType == streamIdxChecksumReq {
			r.processIndexChecksumRequest(req, seeker, seekerResources)
			continue
		}

		reqs = append(reqs, req)
	}

	// Now deal with limit.
	if err := r.queryLimits.AnyExceeded(); err != nil {
		for _, req := range reqs {
			req.onError(err)
		}
		return
	}

	// Now process "reqs" as per normal.

```

Reply:
Sounds good :+1: